go/vt/vttablet/grpctabletconn/conn.go

```diff
@@ -675,18 +675,18 @@ func (conn *gRPCQueryClient) VStreamRows(ctx context.Context, target *querypb.Ta
 		return err
 	}
 	for {
-		r, err := stream.Recv()
+		r := binlogdatapb.VStreamRowsResponseFromVTPool()
+		err := stream.RecvMsg(r)
 		if err != nil {
 			return tabletconn.ErrorFromGRPC(err)
 		}
-		select {
-		case <-ctx.Done():
+		if ctx.Err() != nil {
 			return ctx.Err()
-		default:
 		}
 		if err := send(r); err != nil {
 			return err
 		}
+		r.ReturnToVTPool()
```

Comment:
should the `r.ReturnToVTPool()` not be done in a defer so we don't miss the two other exit points from this method?

Reply:
You cannot `defer` inside of a `for` loop: each `defer` call pushes the cleanup callback to the stack, so you'd overflow very quickly. It's perfectly fine _not_ to return to the pool the message if we don't have a clean exit: not returning objects to the pool is always safe, they get GC'ed all the same. So we can afford to only return the message to the pool in the happy path. :+1: